wallet-new/src/Cardano/Wallet/Kernel/Accounts.hs

```diff
@@ -0,0 +1,208 @@
+                      createHdRndAccount spendingPassword
+                                         accountName
+                                         esk
+                                         hdRootId
+                                         pw
+
+-- | Creates a new 'Account' using the random HD derivation under the hood.
+-- This code follows the same pattern of 'createHdRndAddress', but the two
+-- functions are "similarly different" enough to not make convenient generalise
+-- the code.
+createHdRndAccount :: PassPhrase
+                   -> AccountName
+                   -> EncryptedSecretKey
+                   -> HdRootId
+                   -> PassiveWallet
+                   -> IO (Either CreateAccountError HdAccount)
+createHdRndAccount _spendingPassword accountName _esk rootId pw = do
+    gen <- createSystemRandom
+    go gen 0
+    where
+        go :: GenIO -> Word32 -> IO (Either CreateAccountError HdAccount)
+        go gen collisions =
+            case collisions >= maxAllowedCollisions of
+                 True  -> return $ Left (CreateAccountHdRndAccountSpaceSaturated rootId)
+                 False -> tryGenerateAccount gen collisions
+
+        tryGenerateAccount :: GenIO
+                           -> Word32
+                           -- ^ The current number of collisions
+                           -> IO (Either CreateAccountError HdAccount)
+        tryGenerateAccount gen collisions = do
+            newIndex <- deriveIndex (flip uniformR gen) HdAccountIx HardDerivation
+            let hdAccountId = HdAccountId rootId newIndex
+                newAccount  = initHdAccount hdAccountId genesisCheckpoint &
+                              hdAccountName .~ accountName
+                db = pw ^. wallets
+            res <- update db (CreateHdAccount newAccount)
+            case res of
+                 (Left (CreateHdAccountExists _)) ->
+                     go gen (succ collisions)
+                 (Left (CreateHdAccountUnknownRoot _)) ->
+                     return (Left $ CreateAccountUnknownHdRoot rootId)
+                 Right () -> return (Right newAccount)
+
+        -- The maximum number of allowed collisions.
+        maxAllowedCollisions :: Word32
+        maxAllowedCollisions = 42
+
+        genesisCheckpoint :: Checkpoint
+        genesisCheckpoint = Checkpoint {
+              _checkpointUtxo        = InDb mempty
+            , _checkpointUtxoBalance = InDb (mkCoin 0)
+            , _checkpointPending     = emptyPending
+            , _checkpointBlockMeta   = mempty
+            }
+
+
+-- | Deletes an HD 'Account' from the data storage.
+deleteAccount :: HdAccountId -> PassiveWallet -> IO (Either DeleteAccountError ())
+deleteAccount hdAccountId pw = do
+    res <- liftIO $ update (pw ^. wallets) (DeleteHdAccount hdAccountId)
+    return $ case res of
+         Left (DeleteHdAccountUnknownRoot _unknownRoot) ->
+             Left (DeleteAccountUnknownHdRoot (hdAccountId ^. hdAccountIdParent))
```

Comment:
Why do you ignore the `_unknownRoot` argument here and then reconstruct it yourself?

Reply:
I think just because I was lazy and I didn't want to pattern match on the `UnknownRoot` which in principle has more than just one constructor. Was conveniency really, but I think I can make this a bit more sane.